apollo-router/src/introspection.rs

```diff
@@ -14,33 +14,37 @@ use crate::query_planner::QueryPlanResult;
 const DEFAULT_INTROSPECTION_CACHE_CAPACITY: NonZeroUsize =
     unsafe { NonZeroUsize::new_unchecked(5) };
 
+pub(crate) async fn default_cache_storage() -> CacheStorage<String, Response> {
+    // This cannot fail as redis is not used.
+    CacheStorage::new(DEFAULT_INTROSPECTION_CACHE_CAPACITY, None, "introspection")
+        .await
+        .expect("failed to create cache storage")
+}
+
 /// A cache containing our well known introspection queries.
 pub(crate) struct Introspection {
     cache: CacheStorage<String, Response>,
     pub(crate) planner: Arc<Planner<QueryPlanResult>>,
 }
 
 impl Introspection {
-    pub(crate) async fn with_capacity(
+    pub(crate) async fn with_cache(
         planner: Arc<Planner<QueryPlanResult>>,
-        capacity: NonZeroUsize,
+        cache: CacheStorage<String, Response>,
```

Comment:
I see why we now need to pass the cache as an argument, but I feel it points to something wrong in the `activate` call. Instead of calling it directly everywhere, could the plugins register a closure to handles activation when they are created?

Reply:
I agree with the sentiment of this, but none of the query planner or APQ stuff are plugins, so they don't get to participate lifecycle and these are the point that need to register a closure. We need to discuss the lifecycle hooks that we need and work towards refactoring both non-plugin and plugin code to use an agreed upon model.

As part of this we should also examine what we want in terms of hot reload functionality, as actually having any hot reload significantly complicates our codebase.